### ISSUE TITLE
Tags

### ISSUE THREAD

**Issue by omariom** [PerfDotNet/BenchmarkDotNet#74]
It would be very convenient to be able to attach tags to a benchmark. Each tag should become a column in  the report. This would allow easier sorting, filtering, charting in Excel.

**Comment by AndreyAkinshin** [comment]
Ok, it is very simple to add with help of `ResultExtenders`. What do you think about:
* How it should look? 
* What should we print in the tags columns?
/cc @mattwarren

**Comment by omariom** [comment]
May be like this
```C#
[Benchmark(Tags = new { Kind = "Simple", Type = "int", RangeCheck = "yes"})]
public void SimpleIntWithRangeCheck()

[Benchmark(Tags = new { Kind = "Simple", Type = "long", RangeCheck = "yes"})]
public void SimpleLongWithRangeCheck()

[Benchmark(Tags = new { Kind = "Unrolling", Type = "int", RangeCheck = "no"})]
public void LoopUnrollingIntWithoutRangeCheck()




```

**Comment by AndreyAkinshin** [comment]
@omariom, unfortunately, we can't keep instances of reference types in attributes.

**Comment by omariom** [comment]
![philosoraptor-blank-2](https://cloud.githubusercontent.com/assets/1781701/12379080/e9a2c00a-bd63-11e5-9a30-146d164680c9.jpg)

**Comment by mattwarren** [comment]
Yeah this seems like a nice use-case for ResultExtendes. Although as has been pointed out, we need to decide how the Tags can be expressed in attributes first.

But once that's done, I guess the output could be something like this:

|   Method |     AvrTime |     Error |    Kind |    Type |     RangeCheck |
--------- |------------ |---------- |------ |------ |------ |
 SimpleIntWithRangeCheck| 102.8345 ns | 0.0000 ns | Simple | int | yes |
 SimpleLongWithRangeCheck | 171.6548 ns | 0.0000 ns | Simple | long | yes |
 LoopUnrollingIntWithoutRangeCheck| 207.6533 ns | 0.0000 ns | Unrolling | int | no |

**Comment by AndreyAkinshin** [comment]
I can suggest to keep tags in a string which will be parsed. E.g. `Tags = "Kind=Simple;Type=int;RangeCheck=yes"`. It looks a little ugly, but I think, there is no nice way to do it via attributes.

**Comment by omariom** [comment]
:+1:

**Comment by AndreyAkinshin** [comment]
@omariom, I'm writing new API for BenchmarkDotNet. Here is an implementation of the tags: https://github.com/PerfDotNet/BenchmarkDotNet/tree/f149cb14c864e1b09d2f71e3bf558f9bb59bf77b#columns What do you think? Is it ok for you?

**Comment by omariom** [comment]
@AndreyAkinshin 
The change is big. I need to try it first.

**Comment by omariom** [comment]
Regardless of the change, there is a good rule I believe  every API should follow: "Simple things should be simple, complex things should be possible".

**Comment by mattwarren** [comment]
> there is a good rule I believe every API should follow: "Simple things should be simple, complex things should be possible".

This got me thinking a bit and I'm worried we've lost a bit of simplicity with the code in the "refactoring" branch

Compare the old way of specifying [Benchmarks to run on the CLR and Mono](https://github.com/PerfDotNet/BenchmarkDotNet/blob/8e5a0ed384e8b645e74ca19f1c2fe5e5a921077a/BenchmarkDotNet.Samples/Framework/Framework_StopwatchVsDateTime.cs#L9-L11):

``` csharp
[BenchmarkTask(runtime: BenchmarkRuntime.Clr)]
[BenchmarkTask(runtime: BenchmarkRuntime.Mono)]
public class Framework_StopwatchVsDateTime
```

with the [new way using Config classes](https://github.com/PerfDotNet/BenchmarkDotNet/blob/refactoring/BenchmarkDotNet.Samples/Framework/Framework_DateTime.cs#L9-L16), where you need to declare an entire class (**if I'm missing something and there is already an easier way let me know**):

``` csharp
[Config(typeof(Config))]
public class Framework_DateTime
{
    private class Config : ManualConfig
    {
        public Config()
        {
            Add(Job.Clr, Job.Mono);
            ...
        }
    }
}
```

Whilst I think the new way is great and gives a lot of power and flexibility in more advanced situations, I worried it makes things harder for more simple scenarios?

Can we do something to make simpler scenarios easier to write? For instance is there a way that you could still specify it all in `attributes` for a basic set-up like `CLR` and `Mono` or `x86` and `RyuJIT` and then only drop down to using a `Config` class if you need to create more complex set-ups?

**Comment by AndreyAkinshin** [comment]
For now, you can write something like this:
```cs
[Config("jobs=clr,mono")]
public class Framework_DateTime
```
Looks pretty easy for me. Look at the information about the command style for configs in the `refactoring` README.

**Comment by mattwarren** [comment]
> Looks pretty easy for me. Look at the information about the command style for configs in the `re-factoring` README.

Yeah your right, that's pretty easy. 

Definitely a case of needing to RTFM on my part ;-)

**Comment by adamsitnik** [comment]
I think that configuration with parsable strings might be just too error prone.

Maybe we could combine [Flags] enums with  and Attributes where AllowMultiple = true?

[Some sample code](https://gist.github.com/adamsitnik/c7167a340b1f4be76bbf)

**Comment by mattwarren** [comment]
> I think that configuration with parsable strings might be just too error prone.

Yeah that's a good point, "magic strings" are always tricky

**Comment by omariom** [comment]
Why not preserve (somehow) BenchmarkTask and add BenchmarkTaskConfig?

**Comment by AndreyAkinshin** [comment]
Ok, here are some points:
* New "magic strings" completely match to command line arguments, so, you can copy-paste your config between the `Config` attribute and console. Also, it allows you to set complex config in a short and human-readable form (e.g. `jobs=LegacyJitX64,RyuJitX64 columns=min,max`). You always can copy a desired config from examples in Readme and improve it with a few symbols.
* Old `BenchmarkTask` (aka `Job` after the refactoring) is too complex to develop and extend. For now, we have a `toolchain` that can't be easily added as an attribute property. I want to add a lot of new Job's properties. E.g. I want to add [Runtime Settings](https://msdn.microsoft.com/en-us/library/0kk0kk35.aspx) and it will look ugly as attribute arguments because of CLR metadata capabilities. It is very hard to use predefined jobs (or other kind of BenchmarkDotNet extensions), inherit it, or use several jobs at once ([Flags] enum is not a silver bullet; e.g. you can't ask for `LegacyJitX86,LegacyJitX64,RyuJITx64` at once because RyuJIT doesn't have x86-version). And you can't use static object instances, you should make constants. E.g. I want to add additional [statistics columns](https://github.com/PerfDotNet/BenchmarkDotNet/blob/5df5532c52e26e9e6a21f65457e88067bbb5364e/BenchmarkDotNet/Columns/StatisticColumn.cs#L10) via attributes. How I can possible do it? I should make a separated constants per statistics (min,q1, median, mean, q2, max, confidence intervals and so on) and match each of them to real code. In the refactoring branch, it is easy to define a new column (in the core or as a 3rdParty extensions) with help of an labmda expression and use it everywhere.
* New object style allows you to implement any logic that set you config (e.g. `Tags`) in a nice way. Yes, you should write several lines instead of a single attribute. On the other hand, you can do anything you want. And it is really easy to read and write such configs.
* New "command style" and "object style" are full-featured systems. By design, you can express anything in both styles. Of course, we can add a `Job` attribute as an alternative way to create some configs. But it will be limited, you can't express anything via attributes. Why we should do it? In a simple scenario, we can use "command style" for simple configs (like `jobs=AllJits`). In a complex scenario (when we want to write some additional code), we can use "object style". It is very hard to design limeted-featured additional attributes that will be really useful.

I really want to make the BenchmarkDotNet API simple. I spent a lot of time trying to create a design that would fit for all tasks. New config system is the best design that I have. I wrote benchmarks last two weeks with help of new API, and it is really simple to use it. 

Anyway, I'm ready to discuss any suggestion about alternative way to set configs.

**Comment by omariom** [comment]
@AndreyAkinshin
> Ok, here are some points:

Convinced! )

The last question. Will it be possible to do the following?

```C#
[MyBenchmarkConfig(warmCount: 4)]
class BenchmarkManualSerialization
{
}

public class MyBenchmarkConfigAttribute : ConfigAttribute, IConfig
{
    private IConfig manualConfig = new ManualConfig();
    
    public MyBenchmarkConfigAttribute(int processCount = 3, int warmCount = 3)
        : base(typeof(MyBenchmarkConfigAttribute))
    {
        // Somehow add processCount and warmCount to manualConfig
    }
    
    IEnumerable<string> IConfig.GetColumns()
    {
        return manualConfig.GetColumns();
    }
    
    // Implement the rest of IConfig..
}
```

**Comment by AndreyAkinshin** [comment]
It is an interesting idea. Now we have the following source code:
```cs
[AttributeUsage(AttributeTargets.Class)]
public class ConfigAttribute : Attribute
{
    public IConfig Config { get; }

    public ConfigAttribute(Type type)
    {
        Config = (IConfig)Activator.CreateInstance(type);
    }

    public ConfigAttribute(string command)
    {
        Config = new ConfigParser().Parse(command.Split(' '));
    }
}
```
I can define `IConfigAttribute`:
```cs
    public interface IConfigAttribute
    {
        IConfig Config { get; }
    }

    [AttributeUsage(AttributeTargets.Class)]
    public class ConfigAttribute : Attribute, IConfigAttribute
    {
        // ...
```
So, anyone who implements `IConfigAttribute` can be ConfigAttribute. @omariom, what do you think?

**Comment by omariom** [comment]
@AndreyAkinshin 
:+1: I would just call it IConfigSource.

**Comment by AndreyAkinshin** [comment]
@omariom, done: https://github.com/PerfDotNet/BenchmarkDotNet/blob/120b7c8bb3720c3a48a26840fcec71c30ab5c7b7/BenchmarkDotNet.Samples/Intro/IntroConfigSource.cs

**Comment by omariom** [comment]
:+1: :+1:

**Comment by AndreyAkinshin** [comment]
Implemented in `v0.9.0`.
